src/inc/readytorun.h

```diff
@@ -310,6 +310,12 @@ enum ReadyToRunHelper
     READYTORUN_HELPER_PersonalityRoutineFilterFunclet = 0xF1,
 #endif
 
+    // Synchronized methods
+    READYTORUN_HELPER_MonitorEnter              = 0xF8,
+    READYTORUN_HELPER_MonitorExit               = 0xF9,
+    READYTORUN_HELPER_MonitorEnterStatic        = 0xFA,
```

Comment:
Do the static methods actually work? I would expect them to contain special kind of fixups that will fail with R2R.

Reply:
It compiled and returned expected result for a synchronized Main that returned 123. But I didn't do extensive testing beyond that.

```
Compiling input file C:\git\clr\bin\Product\Windows_NT.x64.Debug\othertest.exe
Compiling method Program.Main
; Assembly listing for method Program:Main():int
; Emitting BLENDED_CODE for X64 CPU with SSE2 - Windows
; debuggable code
; rbp based frame
; fully interruptible
; Final local variable assignments
;
;  V00 OutArgs      [V00    ] (  1,  1   )  lclBlk (32) [rsp+0x00]   "OutgoingArgSpace"
;  V01 tmp1         [V01    ] (  1,  1   )   ubyte  ->  [rbp-0x08]   do-not-enreg[X] addr-exposed "Synchronized method monitor acquired boolean"
;  V02 tmp2         [V02    ] (  1,  1   )     int  ->  [rbp-0x0C]   "fgInsertCommaFormTemp is creating a new local variable"
;  V03 tmp3         [V03    ] (  1,  1   )     int  ->  [rbp-0x10]   "Single return block return value"
;  V04 PSPSym       [V04    ] (  1,  1   )    long  ->  [rbp-0x20]   do-not-enreg[X] addr-exposed "PSPSym"
;
; Lcl frame size = 64

G_M55383_IG01:
       55                   push     rbp
       4883EC40             sub      rsp, 64
       488D6C2440           lea      rbp, [rsp+40H]
       488965E0             mov      qword ptr [rbp-20H], rsp

G_M55383_IG02:
       33C9                 xor      ecx, ecx
       894DF8               mov      dword ptr [rbp-08H], ecx

G_M55383_IG03:
       488B0D00000000       mov      rcx, qword ptr [(reloc)]
       488D55F8             lea      rdx, bword ptr [rbp-08H]
       FF1500000000         call     [CORINFO_HELP_MON_ENTER_STATIC]
       C745F47B000000       mov      dword ptr [rbp-0CH], 123
       488B0D00000000       mov      rcx, qword ptr [(reloc)]
       488D55F8             lea      rdx, bword ptr [rbp-08H]
       FF1500000000         call     [CORINFO_HELP_MON_EXIT_STATIC]
       8B45F4               mov      eax, dword ptr [rbp-0CH]
       8945F0               mov      dword ptr [rbp-10H], eax
       EB00                 jmp      SHORT G_M55383_IG04

G_M55383_IG04:
       8B45F0               mov      eax, dword ptr [rbp-10H]

G_M55383_IG05:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret

G_M55383_IG06:
       55                   push     rbp
       4883EC30             sub      rsp, 48
       488B6920             mov      rbp, qword ptr [rcx+32]
       48896C2420           mov      qword ptr [rsp+20H], rbp
       488D6D40             lea      rbp, [rbp+40H]

G_M55383_IG07:
       488B0D00000000       mov      rcx, qword ptr [(reloc)]
       488D55F8             lea      rdx, bword ptr [rbp-08H]
       FF1500000000         call     [CORINFO_HELP_MON_EXIT_STATIC]
       90                   nop

G_M55383_IG08:
       4883C430             add      rsp, 48
       5D                   pop      rbp
       C3                   ret

; Total bytes of code 119, prolog size 14 for method Program:Main():int
```

Reply:
The `[(reloc)]` above points to `ENCODE_SYNC_LOCK` fixup that is fragile NGen specific. It would need to work to make it work to R2R compatible - give it a stable R2R ID at least. There may be also some more work needed for generic static synchronized methods.

It may be easiest to keep the static synchronized methods on the fallback to JIT plan. They are super rare.